README.md
# LIANNE

***LInk mANagemenNt systEm*** 

This application manage the Bioinformatic workflow for **"Programma di profilazione genomica dei tumori FPG500**

It's designed for the analysis of panel Illumina TruSight Oncology 500 sequenced using the Illumina Novaseq6000 sequencer.

The Variant calling is performed using Illumina Local App and uploaded on PierianDX Clinical Genomics Workspace (CGW) for the variant interpretation.
For these steps is compulsory to have an agreement with Illumina for the Lacal App usage and a licensed account on PierianDX.

It is also use Lianne for coverage analysis using only the analysis coverage jobs 

All sh scripts are designed for PBS scheduler, if it doesn't available on your system the scripts can be used deleting the PBS parameters part



## Requirements

### Conda

Lianne System works in Conda environment, to install Conda follow: 

The conda packages to install are listed in *conda_packages_list.txt*

### Singularity

It also use **Singularity v3.7.4** for the Illumina Local App


### VarHound

The coverage analysis use VarHound available on GitHub at:

`https://github.com/fernandoPalluzzi/VarHound`





## Usage

### First usage

Insert the absolute path of the confPath.ini file in the CONF variable at the beginnign of the lianne.py file.

```
CONF = absolutePath/confPath.ini
```

### Parameters and arguments


```
usage: lianne.py [-h] -i RUNINPUT [-l1 SELECT] [-l2 NCPUS] [-l3 MEM]
                 [-e EMAIL] [-m SENDMODE] [-N NAME] [-q QUEUE]

Link Management System - Lianne

optional arguments:
  -h, --help            show this help message and exit
  -i RUNINPUT, --runInput RUNINPUT
                        NovaSeq output sequencing path
  -l_select SELECT, --select SELECT
                        Select the number of chunks to send on PBS cluser -
                        Default=1
  -l_ncpus NCPUS, --ncpus NCPUS
                        Select the number of ncpus to require - Default=24
  -l_mem MEM, --mem MEM
                        Select the amount of memory to require - Default=128
  -e EMAIL, --email EMAIL
                        Insert the email -
                        Default=luciano.giaco@policlinicogemelli.it
  -m SENDMODE, --sendMode SENDMODE
                        Insert the sending email mode - Default=ae
  -N NAME, --name NAME  Insert the job name - Default=lianne
  -q QUEUE, --queue QUEUE
                        Insert the queue to send job - Default=workq
  -d, --debug           Run the script in debug mode No jobs will be send No
                        file will be written - Default=False
  -f, --fastqc          Perform FastQC analysis on fastq files - Default=False
```

## Steps

Starting Lianne using the default parameters the scrit needs only the path folder of sequencer output.


### Demultiplexing

The demultiplexing results are stored in a temporary folder with the following name *analysis_runName* where runName is the name of sequencing output folder

Lianne performs:

1. A check if the samplesheet exists. If the samplesheet not exists, Lianne returns a warning and exit. If the samplesheet has another name, Lianne makes a copy in the temporary using **SampleSheet.csv as file name**

2. Sends the demultiplexing job using Illumina TruSight Oncology 500 Local App with the following command line:

```
#! /bin/bash 

#PBS -o /temporaryFolder/analysis_runName/stdout_demultiplex
#PBS -e /temporaryFolder/analysis_runName/stderr_demultiplex
#PBS -l select=1:ncpus=24:mem=128g
#PBS -M your@email.com
#PBS -m ae
#PBS -N lianne_demultiplex
#PBS -q workq

module load singularity/3.7.4
module load openmpi/4.1.1
cd /apps/trusight/2.2.0

./TruSight_Oncology_500_RUO.sh \
--analysisFolder /temporaryFolder/analysis_runName/runName \
--resourcesFolder /apps/trusight/2.2.0/resources \
--runFolder /data/novaseq/Diagnostic/NovaSeq/SequencerOutput/runName \
--engine singularity \
--sampleSheet /temporaryFolder/analysis_runName/SampleSheet.csv \
--isNovaSeq \
--demultiplexOnly
```


### Local App

The Illumina TruSigth Oncology Local App v2.2.0 is used for the local variant calling analysis.

Lianne sends in queue the Local App using a sh script containing the following command line and PBS parameters:

```
#! /bin/bash 

#PBS -o /data/novaseq_results/runName/stdout_LocalApp
#PBS -e /data/novaseq_results/runName/stderr_LocalApp
#PBS -l select=2:ncpus=24:mem=128g
#PBS -M your@email.com
#PBS -m ae
#PBS -N lianne_LocalApp
#PBS -q workq

module load singularity/3.7.4
module load openmpi/4.1.1
cd /apps/trusight/2.2.0

./TruSight_Oncology_500_RUO.sh \
--analysisFolder /data/novaseq_results/runName \
--resourcesFolder /apps/trusight/2.2.0/resources \
--runFolder /data/novaseq/Diagnostic/NovaSeq/SequencerOutput/runName \
--engine singularity \
--sampleSheet /temporaryFolder/analysis_runName/SampleSheet.csv \
--isNovaSeq
```

### CGW_runUploader

It needs to have a PiarianDX account

```
#! /bin/bash 

#PBS -o /yourTmp/analysis_220314_A01423_0026_AHGGJNDRXY/stdout_cgwUpload
#PBS -e /yourTmp/analysis_220314_A01423_0026_AHGGJNDRXY/stderr_cgwUpload
#PBS -l select=1:ncpus=5:mem=80g
#PBS -M your@email.com
#PBS -m ae
#PBS -N lianne_cgwUpload
#PBS -q workq

module load corretto/8.292.10.1
cd /PathTo/CGWRunUploader/

java -jar -Dloader.main=com.pdx.commandLine.ApplicationCommandLine RunUploader-1.13.jar --commandLine --runFolder=/yourTmp/analysis_220314_A01423_0026_AHGGJNDRXY/220314_A01423_0026_AHGGJNDRXY --sequencer=Illumina --sequencerFileType=fastq
```

### FastQC

It expects the fastq files in:

`/output_analysis/Logs_Intermediates/FastqGeneration/Sample_ID/sampleID.fastq.gz`



```
#! /bin/bash 

#PBS -o /yourTmp/analysis_220314_A01423_0026_AHGGJNDRXY/stdout_FastQC
#PBS -e /yourTmp/analysis_220314_A01423_0026_AHGGJNDRXY/stderr_FastQC
#PBS -l select=1:ncpus=10:mem=20g
#PBS -M your@email.com
#PBS -m ae
#PBS -N lianne_FastQC
#PBS -q workq



module load anaconda/3
init bash
source ~/.bashrc
conda activate /PathTo/condaEnv/lianne


/PathTo/lianne/Lmodules/fastqc.py -f /yourResults/220314_A01423_0026_AHGGJNDRXY/Logs_Intermediates/FastqGeneration/*/*.fastq.gz
```

### Coverage

It expects the bam file as following:

snv: /output_analysis/Logs_Intermediates/StitchedRealigned/Sample_ID/Sample_ID.bam
cnv: /output_analysis/Logs_Intermediates/DnaRealignment/Sample_ID/Sample_ID.bam
rna: /output_analysis/Logs_Intermediates/RnaMarkDuplicate/Sample_ID/Sample_ID.bam

This analysis is performed using mosdepth package

`https://github.com/brentp/mosdepth`


```
#! /bin/bash 

#PBS -o /yourResults/220314_A01423_0026_AHGGJNDRXY/stdout_cvLaunch
#PBS -e /yourResults/220314_A01423_0026_AHGGJNDRXY/stderr_cvLaunch
#PBS -l select=1:ncpus=1:mem=1g
#PBS -M your@email.com
#PBS -m ae
#PBS -N lianne_cvLaunch
#PBS -q workq



cd /PathTo/lianne/
python3 Lmodules/cvLaunch.py -o /yourResults/220314_A01423_0026_AHGGJNDRXY -e your@email.com -p snv
python3 Lmodules/cvLaunch.py -o /yourResults/220314_A01423_0026_AHGGJNDRXY -e your@email.com -p rna
python3 Lmodules/cvLaunch.py -o /yourResults/220314_A01423_0026_AHGGJNDRXY -e your@email.com -p cnv
```



### VarHound

This analysis is performed using R


```
module load anaconda/3
init bash
source ~/.bashrc
conda activate /PathTo/condaEnv/lianne
cd /yourResults/220314_A01423_0026_AHGGJNDRXY

cd /PathTo/lianne/
python3 VarHound/vhLaunch.py /yourResults/220314_A01423_0026_AHGGJNDRXY/snv_coverage 
python3 VarHound/vhLaunch.py /yourResults/220314_A01423_0026_AHGGJNDRXY/rna_coverage
python3 VarHound/vhLaunch.py /yourResults/220314_A01423_0026_AHGGJNDRXY/cnv_coverage
```






CHANGELOG
######### 2022.03.30

1. Added configuration file

######### 2022.01.14

1. Fixed bug for coverage and varhound jobs
2. Added coverage for rna and cnv
3. added flag for FastQC analysis

######### 2022.01.14

1. Fixed bug in normpath for runInput path

######### 2022.01.12

1. Added VarHound for coverage analysis

######### 2021.11.29

1. Changed queue for FastQC


######### 2021.11.19

Refined the workflow and job queue
Step available:
1. Demultiplexing
2. Fastq upload to CGW using CGWRunUploader
3. Fastq QC using FastQC
4. Analysis using Illumina TruSigt Local App v2.2
5. Making sequence details csv file to run jobs on CGW
6. Added debug mode


######### 2021.10.21

Added Mosdepth to calculate coverage at 50x 100x 150x 500x
Ref.
Brent S Pedersen, Aaron R Quinlan, Mosdepth: quick coverage calculation for genomes and exomes, Bioinformatics, Volume 34, Issue 5, 01 March 2018, Pages 867–868, https://doi.org/10.1093/bioinformatics/btx699

lianne.py


#####################################
# NAME: lianne.py
# AUTHOR: Luciano Giaco'
# Date: 06/07/2021
version = "1.0"
# ===================================


import os
import argparse
import subprocess
import make_seq_details
from configparser import ConfigParser
from shutil import copyfile

CONF = '/Users/lucianogiaco/lianne/conf/confPath.ini'

# instantiate
config = ConfigParser()

# parse existing file
# read config file
configFile = config.read(CONF)
RESULTS = config.get('path', 'RESULTS')
TMP = config.get('path', 'TMP')
LOCAL_APP = config.get('path', 'LOCAL_APP')
D_RESOUCES = config.get('path', 'D_RESOUCES')
LIANNE_FOLDER = config.get('path', 'LIANNE_FOLDER')
CGWRunUploader = config.get('path', 'CGWRunUploader')
lianne_env = config.get('path', 'lianne_env')
COV_MODULE = os.path.join(LIANNE_FOLDER, 'Lmodules/coverage.py')



#####################################
# Classes
# ===================================

class pbs_parameters:

    def __init__(self, pathStd, select, ncpus, mem, email, sendMode, name, queue, m):

        # ARGUMENTS
        # 
        # pathStd = Folder path where stdin and stderr files will be created - String
        # select = Numebr of chunks request for PBS - Integer
        # ncpus = Number of cpus request for PBS - Integer
        # mem = Amount of RAM request for PBS in Gb - Integer
        # email = email adress used by PBS to send the email - String
        # sendMode = Send parameter mode 'abe' - String
        # name = Job name - String
        # queue = Queue to use on cluster - workq or gpuq 

        self.pathStdout = os.path.join(pathStd, 'stdout_'+m)
        self.pathStderr = os.path.join(pathStd, 'stderr_'+m)
        self.resources = 'select='+str(select)+':ncpus='+str(ncpus)+':mem='+str(mem)
        self.email = email
        self.sendMode = sendMode
        self.name = name+'_'+m
        self.queue = queue

    def getStdout(self):
        return self.pathStdout

    def getStderr(self):
        return self.pathStderr    

    def getResources(self):
        return self.resources

    def getEmail(self):
        return self.email

    def getMode(self):
        return self.sendMode

    def getName(self):
        return self.name

    def getQueue(self):
        return self.queue
            

#####################################
# Functions
# ===================================

def get_folderOut(runInput):
	pathInput = os.path.normpath(runInput)
	head, tail = os.path.split(pathInput)
	return(tail)

def build_param_sh(parameters):

	o = parameters.getStdout()
	e = parameters.getStderr()
	l = parameters.getResources()
	M = parameters.getEmail()
	m = parameters.getMode()
	N = parameters.getName()
	q = parameters.getQueue()

	
	par = '#! /bin/bash \n\n'
	par = par+'#PBS -o '+o+'\n'
	par = par+'#PBS -e '+e+'\n'
	par = par+'#PBS -l '+l+'\n'
	par = par+'#PBS -M '+M+'\n'
	par = par+'#PBS -m '+m+'\n'
	par = par+'#PBS -N '+N+'\n'
	par = par+'#PBS -q '+q+'\n\n'

	return(par)

def demultiplex_cl(tmp_fastq, runInput, samplesheet):
	# demultiplex command line
	
	d_cl = 'module load singularity/3.7.4\n'
	d_cl = d_cl+'module load openmpi/4.1.1\n'
	d_cl = d_cl+'cd '+LOCAL_APP+'\n\n'
	d_cl = d_cl+'./TruSight_Oncology_500_RUO.sh '
	d_cl = d_cl+'--analysisFolder '
	d_cl = d_cl+tmp_fastq+' '
	d_cl = d_cl+'--resourcesFolder '
	d_cl = d_cl+D_RESOUCES+' '
	d_cl = d_cl+'--runFolder '
	d_cl = d_cl+runInput+' '
	d_cl = d_cl+'--engine singularity '
	d_cl = d_cl+'--sampleSheet '
	d_cl = d_cl+samplesheet+' '
	d_cl = d_cl+'--isNovaSeq '
	d_cl = d_cl+'--demultiplexOnly'

	return d_cl

def samplesheet_manage(samplesheet, tmp_path, debug):
	if os.path.isfile(samplesheet):
		return [True]
	else:
		# run folder list
		dirlist = os.listdir(runInput)
		# extract csv files
		matching = [s for s in dirlist if ".csv" in s]
		# control if there are more then one csv file
		# if not, copy the csv found with SampleSheet.csv name
		if len(matching) != 1:
			print('# WARNING - More than one csv file found\n# INFO exit')
			os.sys.exit()
		else:
			src = os.path.join(runInput, matching[0])
			samplesheet = os.path.join(tmp_path, 'SampleSheet.csv')
			if debug is False:
				copyfile(src, samplesheet)
		return [False, samplesheet]

def localApp_cl(out_localApp, runInput, samplesheet):
	dr_cl = 'module load singularity/3.7.4\n'
	dr_cl = dr_cl+'module load openmpi/4.1.1\n'
	dr_cl = dr_cl+'cd '+LOCAL_APP+'\n\n'
	dr_cl = dr_cl+'./TruSight_Oncology_500_RUO.sh '
	dr_cl = dr_cl+'--analysisFolder '
	dr_cl = dr_cl+out_localApp+' '
	dr_cl = dr_cl+'--resourcesFolder '
	dr_cl = dr_cl+D_RESOUCES+' '
	dr_cl = dr_cl+'--runFolder '
	dr_cl = dr_cl+runInput+' '
	dr_cl = dr_cl+'--engine singularity '
	dr_cl = dr_cl+'--sampleSheet '
	dr_cl = dr_cl+samplesheet+' '
	dr_cl = dr_cl+'--isNovaSeq'

	return(dr_cl)


####################
##  --> MAIN <--  ##
####################

def main(runInput, select, ncpus, mem, email, sendMode, name, queue, debug, fastqc):


	tail = get_folderOut(runInput)
	print('TAIL')
	print(tail)
	out_localApp = os.path.join(RESULTS, tail)
	if os.path.exists(out_localApp):
		print('[INFO] path results already exists')
		print('[INFO] check the path: '+out_localApp)
		print('[INFO] exit')
		os.sys.exit()
	################
	# DEMULTIPLEXING

	# path management
	# analysis folder will created
	# eg. analysis_210729_A01423_0009_AH33WGDRXY
	tmp_path = os.path.join(TMP, 'analysis_'+tail)
	tmp_path = os.path.normpath(tmp_path)

	if debug is False:
		try:
			os.mkdir(tmp_path)
		except FileExistsError:
			print('[WARNING] directory '+tmp_path+' already exists')
			pass
	
	# path folder used in d_file as --analysisFolder parameter
	tmp_fastq = os.path.join(tmp_path, tail)
	if debug is True:
		print('[DEBUG] path folder used in d_file as --analysisFolder parameter')
		print(tmp_fastq)
		print('\n')
	

	# sample sheet check
	samplesheet = os.path.join(runInput, 'SampleSheet.csv')
	control = samplesheet_manage(samplesheet, tmp_path, debug)
	
	if control[0] == False:
		print('# INFO - Found a SampleSheet with different name\n# INFO - '+control[1]+' copied')
		samplesheet = control[1]
	
	
	
	

################################
### QSUB JOBS

	################
	# Demultiplexing

	# pbs parameters
	parameters = pbs_parameters(tmp_path, select, ncpus, mem, email, sendMode, name, queue, 'demultiplex')
	par = build_param_sh(parameters)

	# command line
	d_cl = demultiplex_cl(tmp_fastq, runInput, samplesheet)
	d_sh = par+d_cl

	## DEBUG
	if debug is True:
		print('[DEBUG] Demultiplexing PBS parameters')
		print(par)
		print('\n')

	# build sh file
	d_file = os.path.join(tmp_path, 'demultiplex.sh')

	## DEBUG
	if debug is False:
		sh = open(d_file, 'w')
		sh.write(d_sh)
		sh.close()
	
		# Demultiplexing job
		print('[INFO] Sending '+d_file)
	
		# Capture the job ID for qsub hold
		jobid1 = subprocess.run(['qsub', d_file], stdout=subprocess.PIPE, universal_newlines=True)
		jobid1_str = jobid1.stdout
		
	else:
		print('[DEBUG] sh file written in foder: ')
		print(d_file)
		print('[DEBUG] sh file contains: ')
		print(d_sh)
		print('\n\n\n')

	
	
	################
	# localApp

	# path management
	
	
	select = 2
	pathStd = pbs_parameters(out_localApp, select, ncpus, mem, email, sendMode, name, queue, 'LocalApp')
	par = build_param_sh(pathStd)

	## DEBUG
	if debug is True:
		print('[DEBUG] Local App PBS parameters')
		print(par)
		print('\n')

	# path local app sh file
	dr_file = os.path.join(tmp_path, 'localApp.sh')

	# local app command line
	dr_cl = localApp_cl(out_localApp, runInput, samplesheet)

	# sh file to send
	dr_sh = par+dr_cl

	if debug is False:
		# build sh file
		sh = open(dr_file, 'w')
		sh.write(dr_sh)
		sh.close()
		print('[INFO] Sending '+dr_file)
		# send job
		jobid2 = subprocess.run(['qsub', dr_file], stdout=subprocess.PIPE, universal_newlines=True)
		jobid2_str = jobid2.stdout
	else:
		print('[DEBUG] localApp file written in foder: ')
		print(dr_file)
		print('[DEBUG] localApp file contains:')
		print(dr_sh)
		print('[DEBUG] output folder of local app:')
		print(out_localApp)



	################
	# Upload to CGW
	

	# pbs parameters
	select = 1
	ncpus = 5
	mem = '80g'
	parameters = pbs_parameters(tmp_path, select, ncpus, mem, email, sendMode, name, queue, 'cgwUpload')
	par = build_param_sh(parameters)

	## DEBUG
	if debug is True:
		print('[DEBUG] CGWRunUploader PBS parameters')
		print(par)
		print('\n')


	dr_cl = 'module load corretto/8.292.10.1\n'
	dr_cl = dr_cl+'cd '+CGWRunUploader+'\n'
	dr_cl = dr_cl+'\n'
	dr_cl = dr_cl+'java -jar '
	dr_cl = dr_cl+'-Dloader.main=com.pdx.commandLine.ApplicationCommandLine RunUploader-1.13.jar '
	dr_cl = dr_cl+'--commandLine '
	dr_cl = dr_cl+'--runFolder='+tmp_fastq+' '
	dr_cl = dr_cl+'--sequencer=Illumina '
	dr_cl = dr_cl+'--sequencerFileType=fastq'

	dr_sh = par+dr_cl
	# path management
	cgw_file = os.path.join(tmp_path, 'cgw_uploader.sh')

	if debug is False:
		# build sh file
		sh = open(cgw_file, 'w')
		sh.write(dr_sh)
		sh.close()
		print('[INFO] Sending '+cgw_file)
		dependencyID = 'depend=afterany:'+jobid1_str
		jobid2 = subprocess.run(['qsub', '-W', dependencyID, cgw_file], stdout=subprocess.PIPE, universal_newlines=True)
	else:
		print('[DEBUG] cgw_uploader.sh file written in foder: ')
		print(cgw_file)
		print('[DEBUG] cgw_uploader.sh file contains:')
		print(dr_sh)
		print('\n')
		
	

	###############
	# FastQC

	if fastqc is True:
		# pbs parameters
		select = 1
		ncpus = 10
		mem = '20g'
		parameters = pbs_parameters(tmp_path, select, ncpus, mem, email, sendMode, name, queue, 'FastQC')
		par = build_param_sh(parameters)

		dr_cl = 'module load anaconda/3\n'
		dr_cl = dr_cl+'init bash\n'
		dr_cl = dr_cl+'source ~/.bashrc\n'
		dr_cl = dr_cl+'conda activate '+lianne_env+'\n'
		dr_cl = dr_cl+'\n'
		dr_cl = dr_cl+'\n'
		
		# Set folder where Fastq are located
		# in the Illumina local app output folder
		# and append all FastQC call
		fastq_folder = os.path.join(out_localApp, 'Logs_Intermediates/FastqGeneration/*/*.fastq.gz')
		
		fastqc_path = os.path.join(LIANNE_FOLDER, 'Lmodules/fastqc.py')
		sh_cmd = fastqc_path+' -f '+fastq_folder#+' -t '+tmp_path

		dr_sh = par+'\n\n'+dr_cl+sh_cmd
		FastQC_file_run = os.path.join(tmp_path, 'FastQC_run.sh')
		

		if debug is False:
			# build sh file
			sh = open(FastQC_file_run, 'w')
			sh.write(dr_sh)
			sh.close()
			dependencyID = 'depend=afterany:'+jobid2_str
			jobid3 = subprocess.run(['qsub', '-W', dependencyID, FastQC_file_run], stdout=subprocess.PIPE, universal_newlines=True)
		else:
			print('[DEBUG] FastQC.sh file written in foder: ')
			print(FastQC_file_run)
			print('[DEBUG] FastQC.sh file contains:')
			print(dr_sh)
			print(par)
	else:
		pass


	

	###############
	# Coverage

	# pbs parameters
	select = 1
	ncpus = 5
	mem = '10g'

	pathStd = pbs_parameters(out_localApp, select, ncpus, mem, email, sendMode, name, queue, 'coverage')
	par = build_param_sh(pathStd)

	dr_cl = 'module load anaconda/3\n'
	dr_cl = dr_cl+'init bash\n'
	dr_cl = dr_cl+'source ~/.bashrc\n'
	dr_cl = dr_cl+'conda activate '+lianne_env+'\n'
	dr_cl = dr_cl+'cd '+out_localApp
	dr_cl = dr_cl+'\n'
	dr_cl = dr_cl+'\n'

	dr_sh = par+'\n\n'+dr_cl



	# write coverage sh
	# pbs parameters
	select = 1
	ncpus = 1
	mem = '1g'

	# command line inside the sh job
	pathStd = pbs_parameters(out_localApp, select, ncpus, mem, email, sendMode, name, queue, 'cvLaunch')
	par = build_param_sh(pathStd)
	cv_sh = par+'\n\n'
	cv_sh = cv_sh+'cd '+LIANNE_FOLDER+'\n'
	cv_sh = cv_sh+'python3 Lmodules/cvLaunch.py -o '+out_localApp+' -e '+email+' -p snv\n'
	cv_sh = cv_sh+'python3 Lmodules/cvLaunch.py -o '+out_localApp+' -e '+email+' -p rna\n'
	cv_sh = cv_sh+'python3 Lmodules/cvLaunch.py -o '+out_localApp+' -e '+email+' -p cnv\n'

	cvLaunch = os.path.join(tmp_path, 'cvLaunch.sh')


	if debug is False:
		sh = open(cvLaunch, 'w')
		sh.write(cv_sh)
		sh.close()
		print('[INFO] Sending '+cvLaunch)
		dependencyID = 'depend=afterany:'+jobid2_str
		jobid4 = subprocess.run(['qsub', '-W', dependencyID, cvLaunch], stdout=subprocess.PIPE, universal_newlines=True)
		jobid4_str = jobid4.stdout
	else:
		print('[DEBUG] cvLaunch.sh file written in foder: ')
		print(cvLaunch)
		print('[DEBUG] cvLaunch.sh file contains:')
		print(cv_sh)
		print("\n\n")
		print('[DEBUG] coverage sh file:')
		jobid5 = subprocess.run(['python3', '/data/hpc-data/shared/pipelines/lianne/Lmodules/cvLaunch.py', '-o', out_localApp, '-d'], stdout=subprocess.PIPE, universal_newlines=True)
		jobid5_str = jobid5.stdout
		print(jobid5_str)
	

	###############
	# VarHound

	# pbs parameters
	select = 1
	ncpus = 2
	mem = '5g'

	pathStd = pbs_parameters(out_localApp, select, ncpus, mem, email, sendMode, name, queue, 'lianne_vh')
	par = build_param_sh(pathStd)

	coverage_out_snv = os.path.join(out_localApp, 'snv_coverage')
	coverage_out_rna = os.path.join(out_localApp, 'rna_coverage')
	coverage_out_cnv = os.path.join(out_localApp, 'cnv_coverage')
	dr_cl = 'module load anaconda/3\n'
	dr_cl = dr_cl+'init bash\n'
	dr_cl = dr_cl+'source ~/.bashrc\n'
	dr_cl = dr_cl+'conda activate '+lianne_env+'\n'
	dr_cl = dr_cl+'cd '+out_localApp
	dr_cl = dr_cl+'\n'
	dr_cl = dr_cl+'\n'
	dr_cl = dr_cl+'cd '+LIANNE_FOLDER+'\n'
	dr_cl = dr_cl+'python3 VarHound/vhLaunch.py '+coverage_out_snv+' \n'
	dr_cl = dr_cl+'python3 VarHound/vhLaunch.py '+coverage_out_rna+'\n'
	dr_cl = dr_cl+'python3 VarHound/vhLaunch.py '+coverage_out_cnv+'\n'



	varhound_file_run = os.path.join(tmp_path, 'varhound_run.sh')

	if debug is False:
		sh = open(varhound_file_run, 'w')
		sh.write(dr_cl)
		sh.close()
		print('[INFO] Sending '+varhound_file_run)
		dependencyID = 'depend=afterany:'+jobid4_str
		jobid5 = subprocess.run(['qsub', '-W', dependencyID, varhound_file_run], stdout=subprocess.PIPE, universal_newlines=True)
		jobid5_str = jobid5.stdout
	else:
		print('[DEBUG] varhound_run.sh file written in foder: ')
		print(varhound_file_run)
		print('[DEBUG] varhound_run.sh file contains:')
		print(dr_cl)


	if debug is False:
		print('[INFO] Queue:')
		subprocess.run(['qstat'])
		os.sys.exit()
	else:
		print('[INFO] exit')
		os.sys.exit()

	
	
	
	################
	# BUILD CSV 
	print(samplesheet)
	# 
	# seq details
	
	make_seq_details.main(samplesheet)




if __name__ == '__main__':
	# parser variable
	parser = argparse.ArgumentParser(description='Link Management System - Lianne')

	# arguments
	parser.add_argument('-i', '--runInput', required=True,
						help='NovaSeq output sequencing path')
	parser.add_argument('-l_select', '--select', required=False,
						default = 1,
						help='Select the number of chunks to send on PBS cluser - Default=1')
	parser.add_argument('-l_ncpus', '--ncpus', required=False,
						default = 24,
						help='Select the number of ncpus to require - Default=24')
	parser.add_argument('-l_mem', '--mem', required=False,
						default = '128g',
						help='Select the amount of memory to require - Default=128')
	parser.add_argument('-e', '--email', required=False,
						default = 'your@email.com',
						help='Insert the email - Default = your@email.com')
	parser.add_argument('-m', '--sendMode', required=False,
						default = 'ae',
						help='Insert the sending email mode - Default=ae')
	parser.add_argument('-N', '--name', required=False,
						default = 'lianne',
						help='Insert the job name - Default=lianne')
	parser.add_argument('-q', '--queue', required=False,
						default = 'workq',
						help='Insert the queue to send job - Default=workq')
	parser.add_argument('-d', '--debug', required=False,
						action='store_true',
						help='Run the script in debug mode\nNo jobs will be send\nNo file will be written - Default=False')
	parser.add_argument('-f', '--fastqc', required=False,
						action='store_false',
						help='Perform FastQC analysis on fastq files - Default=False')


	args = parser.parse_args()
	runInput = args.runInput
	select = args.select
	ncpus = args.ncpus
	mem = args.mem
	email = args.email
	sendMode = args.sendMode
	name = args.name
	queue = args.queue
	debug = args.debug
	fastqc = args.fastqc

	main(runInput, select, ncpus, mem, email, sendMode, name, queue, debug, fastqc)